### DISCUSSION TITLE
Slow when start the share

### DISCUSSION THREAD

**Discussion by Bocholand** [excalidraw/excalidraw#3150] [Q&A]
Hi,

I use Exclalidraw in the docker. I don't change the config and I haven't problem of network or hardware.

When I use alone, I haven't problem.
When I share the whiteboard, I haven't problem but the other user watch my actions with slowness.
It can't use whiteboard with share due this slowness.

Do you have a solution ?

**Comment by harshagarwalnyu** [comment]
Slowness in collaborative mode (sharing) on self-hosted Excalidraw is almost always a **WebSocket relay bottleneck**, not hardware. The shared whiteboard routes all drawing operations through the `excalidraw-room` WebSocket server — the receiving client re-renders on every incoming event.

**Diagnose first — open browser DevTools → Network → WS tab:**
- Look at the WebSocket frames when you draw
- If frames are arriving but the rendering is slow: client-side CPU (less likely on modern hardware)
- If frames are delayed: the room server is the bottleneck

**Common causes and fixes:**

**1. excalidraw-room running without persistence (event replay on reconnect)**
By default, `excalidraw-room` buffers the entire room state in memory. Large whiteboards with many elements = large state to broadcast on join/reconnect.
```bash
# Check your room server memory usage
docker stats <excalidraw-room-container>
```
Fix: increase room server memory, or use the Redis adapter for state persistence.

**2. No compression on WebSocket messages**
Excalidraw's room server supports gzip compression. Ensure your reverse proxy (nginx) doesn't disable WebSocket compression:
```nginx
location /socket.io/ {
    proxy_pass http://excalidraw-room:80;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    # Don't decompress — let the client handle it
}
```

**3. Single-threaded room server under load**
`excalidraw-room` is Node.js (single-threaded). Multiple rooms on one instance → contention. Run one instance per room, or use a load balancer with sticky sessions.

**4. Geographic latency**
If the other user is in a different region, RTT adds perceptible delay. Host the room server closer to both users, or use a CDN with WebSocket support.

**Quick test:** have both users join from the same LAN (connect to localhost). If lag disappears, the problem is network latency or server load, not the Excalidraw code.

**Comment by Momoru2002** [comment]
Hi, I checked the excalidraw-room source code directly (the server used for the collaboration/sharing feature), and it looks like some of the explanation in the previous comment isn't quite accurate compared to the actual code. So let me walk through what's actually in there.

The excalidraw-room server is fully stateless - it just relays the "server-broadcast" and "server-volatile-broadcast" events to the other sockets in the same room. There's no in-memory state buffering, no Redis, and no gzip compression configuration anywhere in the code. The data being sent is also already end-to-end encrypted (an encrypted ArrayBuffer), so the server is just forwarding an opaque blob without doing anything to it.

So the most likely causes of the lag you're seeing are one of these:

1. Your reverse proxy (e.g. nginx) isn't configured correctly to upgrade the connection to WebSocket. If the upgrade fails, Socket.IO falls back to HTTP long-polling, which is much slower for frequent updates like drawing. Check your browser DevTools > Network > WS filter, and see whether the connection is actually "websocket" or stuck on "polling".

2. If the drawing is already complex (lots of elements), the watching client (not the one drawing) has to decrypt and re-render on every incoming broadcast from other users. This load is on the client side, not the server, so the more elements on the canvas, the heavier it gets for the viewer's browser.

3. Geographic latency, if you and the other user are far from the room server, also adds noticeable delay.

Try checking the Network tab in DevTools while sharing - if the connection is stuck on polling instead of websocket, that's most likely the root cause, usually from the reverse proxy not being configured correctly for the WebSocket upgrade headers.
